cmake/install.cmake

```diff
@@ -0,0 +1,48 @@
+function(install_headers_from_dir srcdir)
+    #Note: install(DIRECTORY ...) causes empty directories being created
+    get_filename_component(cname "${srcdir}" NAME)
+    foreach (subdir ${ARGN})
+        set(abssrcdir "${CMAKE_SOURCE_DIR}/../${srcdir}/${subdir}")
+        file(GLOB_RECURSE files RELATIVE "${abssrcdir}" "${abssrcdir}/*.h")
+        foreach (f ${files})
+            get_filename_component(d "${f}" DIRECTORY)
+            install(FILES "${abssrcdir}/${f}" DESTINATION "include/onnxruntime/${cname}/${subdir}/${d}")
+        endforeach()
+    endforeach()
+endfunction()
+
+install_headers_from_dir(include/onnxruntime/core platform)
+install_headers_from_dir(onnxruntime/core
+        common
+        framework
+        platform
+        session
+        util
+        providers/shared_library
+        providers/shared
+        providers/cpu
+        )
+
+install_headers_from_dir(orttraining/orttraining
+        models/runner
+        training_ops/cpu
+        core/framework
+        core/graph
+        core/optimizer
+        core/session
+        )
+
+install(FILES
+        ${CMAKE_SOURCE_DIR}/../include/onnxruntime/core/providers/providers.h
+        ${CMAKE_SOURCE_DIR}/../onnxruntime/core/providers/op_kernel_type_control.h
+        ${CMAKE_SOURCE_DIR}/../onnxruntime/core/framework/op_kernel_type_control_utils.h
+        ${CMAKE_SOURCE_DIR}/../onnxruntime/core/providers/common.h
+        ${CMAKE_SOURCE_DIR}/../onnxruntime/core/providers/utils.h
+        ${CMAKE_SOURCE_DIR}/../onnxruntime/core/providers/get_execution_providers.h
+
+        DESTINATION include/onnxruntime/core/providers
+)
+
+# Build Generated Headers
+install(FILES ${CMAKE_CURRENT_BINARY_DIR}/onnxruntime_config.h
```

Comment:
The "onnxruntime_config.h" is generated for the build machine, based on its compiler's capability.  When you use it on a different machine, it might not match.

Reply:
I think there are two options:
1. Make it optional and do not install it. 
2. Install it, and remove the HAVE_CONFIG_H macro, and make sure every C/C++ header file includes this file, and put all ORT specific #ifdef macros in. For example, USE_CUDA, ENABLE_CPU_FP16_TRAINING_OPS/DISABLE_ABSEIL/PLATFORM_POSIX/EIGEN_USE_THREADS/...

I prefer the second, because without the information it's hard to make the compile flags consistent.

Reply:
I found an interesting discussion: https://lists.gnu.org/archive/html/autoconf/2012-09/msg00010.html

Probably the best solution is to have two header files:
1. one is generated by gathering information from the build environment,  and is optional.  The main purpose of this file probably is just for suppressing warnings.
2. Another one is generated based on cmake options. For example, onnxruntime_USE_CUDA. This one must be included by every header file.